libuavcan/test/util/avl_tree.cpp

```diff
@@ -0,0 +1,610 @@
+/*
+ * Copyright (C) 2019 Theodoros Ntakouris <zarkopafilis@gmail.com>
+ */
+
+#include <gtest/gtest.h>
+#include <uavcan/util/avl_tree.hpp>
+#include <uavcan/dynamic_memory.hpp>
+#include <uavcan/driver/system_clock.hpp>
+
+using uavcan::AvlTree;
+
+struct Entry {
+    int key;
+    int payload;
+
+    bool operator<(const Entry& other) const {
+        return this->key < other.key;
+    }
+
+    bool operator>(const Entry& other) const {
+        return this->key > other.key;
+    }
+
+    bool operator==(const Entry& other) const {
+        return this->key == other.key;
+    }
+};
+
+// OOM-Unsafe
+inline Entry* makeEntry(uavcan::PoolAllocator<64 * 24, 64>* allocator, int key, int payload) {
+    void* praw = allocator->allocate(sizeof(Entry));
+
+    Entry* e = new (praw) Entry();
+    UAVCAN_ASSERT(e);
+
+    e->key = key;
+    e->payload = payload;
+    return e;
+}
+
+inline bool matchPostOrder(Entry* expected[], AvlTree<Entry>* tree) {
+    int count = 0;
+    bool res = true;
+
+    tree->walkPostOrder([expected, &count, &res](Entry*& in) {
```

Comment:
this frankly concerns me. We're relying on a given std::function implementation to have a small buffer optimization large enough to hold at least 3 machine words of data for the capture. This seems reasonable but I have no data to know that. Perhaps we should just define a functor we use to do this traversal? Also, what are we using this method for exactly? Only testing?

Reply:
This is on the tests, where heap allocation is permitted. In general, the traverse is only used for tests. It's the only proper way of checking the contents of the tree, without exposing `Node` and `root_`